### PR TITLE
CDAP-13548 Detect cycles in the lineage graph.

### PR DESCRIPTION
**Pull request by sagarkapare**
JIRA: https://issues.cask.co/browse/CDAP-13548
Moved the logic of topological sorting to the FieldLineageInfo as static method and updated it to determine the cycles in the operations as well.

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 1]
```diff
@@ -92,22 +90,48 @@
    * validate the existence of path to the fields in the destination from sources. If no such path exists
    * then the lineage will be incomplete.
    *
+   * Apart from collection of operations, this instance also stores the fields belonging
+   * to the destination along with the incoming and outgoing summaries. Computing summaries
+   * can be expensive operation, so it it better to do it while emitting the instance to TMS
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: nit: it it -> it is

**Review by albertshau — APPROVED** [review]
minor javadoc comment, otherwise lgtm

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 1 above)
Comment (sagarkapare) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: fixed.

**Comment by sagarkapare** [comment]
Build is running here - https://builds.cask.co/browse/CDAP-RUT1562-3/

**Comment by sagarkapare** [comment]
Build is green here - https://builds.cask.co/browse/CDAP-RUT1562-4. Squashing and merging.

**Pull request merged**